src/msw/thread.cpp

```diff
@@ -500,34 +504,50 @@ class wxThreadKeepAlive
 /* static */
 void wxThreadInternal::DoThreadOnExit(wxThread *thread)
 {
+    if ( wxSystemOptions::IsFalse("catch-unhandled-exceptions") )
+    {
+        thread->OnExit();
+    }
+    else
     wxTRY
     {
         thread->OnExit();
     }
     wxCATCH_ALL( wxTheApp->OnUnhandledException(); )
 }
 
+/* static */
+THREAD_RETVAL wxThreadInternal::DoThreadStartWithoutExceptionHandling(wxThread* thread)
+{
+    // store the thread object in the TLS
+    wxASSERT_MSG(gs_tlsThisThread != TLS_OUT_OF_INDEXES,
+        "TLS index not set. Is wx initialized?");
+
+    if (!::TlsSetValue(gs_tlsThisThread, thread))
+    {
+        wxLogSysError(_("Cannot start thread: error writing TLS."));
+
+        return THREAD_ERROR_EXIT;
+    }
+
+    return wxPtrToUInt(thread->Entry());
+}
+
 /* static */
 THREAD_RETVAL wxThreadInternal::DoThreadStart(wxThread *thread)
 {
     wxON_BLOCK_EXIT1(DoThreadOnExit, thread);
 
     THREAD_RETVAL rc = THREAD_ERROR_EXIT;
 
+    if ( wxSystemOptions::IsFalse("catch-unhandled-exceptions") )
+    {
+        rc = DoThreadStartWithoutExceptionHandling(thread);
```

Comment:
```suggestion
        return DoThreadStartWithoutExceptionHandling(thread);
```

and remove `else` below (and move this above `rc` declaration).